src/java/com/rapleaf/jack/queries/GenericQuery.java

```diff
@@ -0,0 +1,149 @@
+package com.rapleaf.jack.queries;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import com.rapleaf.jack.BaseDatabaseConnection;
+import com.rapleaf.jack.ModelField;
+import com.rapleaf.jack.ModelWithId;
+
+public class GenericQuery {
+
+  private final BaseDatabaseConnection dbConnection;
+  private Class<? extends ModelWithId> mainModel;
+  private List<JoinCondition> joinConditions;
+  private List<WhereConstraint> whereConstraints;
+  private List<OrderCriterion> orderCriteria;
+  private Set<ModelField> selectedIModelFields;
+  private Optional<LimitCriterion> limitCondition;
+
+  private GenericQuery(BaseDatabaseConnection dbConnection) {
+    this.dbConnection = dbConnection;
+    this.mainModel = null;
+    this.joinConditions = Lists.newArrayList();
+    this.whereConstraints = Lists.newArrayList();
+    this.orderCriteria = Lists.newArrayList();
+    this.selectedIModelFields = Sets.newHashSet();
+    this.limitCondition = Optional.absent();
+  }
+
+  public static GenericQuery create(BaseDatabaseConnection dbConnection) {
+    return new GenericQuery(dbConnection);
+  }
+
+  public GenericQueryBuilder from(Class<? extends ModelWithId> model) {
+    this.mainModel = model;
+    return new GenericQueryBuilder(dbConnection, this);
+  }
+
+  List<WhereConstraint> getWhereConstraints() {
+    return whereConstraints;
+  }
+
+  Set<ModelField> getSelectedModelFields() {
+    return selectedIModelFields;
+  }
+
+  void addJoinCondition(JoinCondition joinCondition) {
+    this.joinConditions.add(joinCondition);
+  }
+
+  void addWhereCondition(WhereConstraint whereConstraint) {
+    this.whereConstraints.add(whereConstraint);
+  }
+
+  void addOrderCondition(OrderCriterion orderCriterion) {
+    this.orderCriteria.add(orderCriterion);
+  }
+
+  void addLimitCondition(LimitCriterion limitCriterion) {
+    this.limitCondition = Optional.of(limitCriterion);
+  }
+
+  void addSelectedModelField(ModelField modelField) {
+    this.selectedIModelFields.add(modelField);
+  }
+
+  boolean isOrderedQuery() {
+    return !orderCriteria.isEmpty();
+  }
+
+  String getSqlStatement() {
+    StringBuilder statement = new StringBuilder();
+
+    statement.append(getSelectClause())
+        .append(getJoinClause())
+        .append(getWhereClause());
+
+    if (isOrderedQuery()) {
+      statement.append(getOrderClause());
+      if (!orderCriteria.isEmpty()) {
+        statement.append(getLimitClause());
```

Comment:
Why is limit only used when it's an ordered query? I may want to limit a query without specifying an order.
Also, aren't isOrderedQuery() and "if (!orderCriteria.isEmpty()) {" redundant of each other?